salt/modules/macdefaults.py

```diff
@@ -98,15 +223,275 @@ def delete(domain, key, user=None):
 
         salt '*' macdefaults.delete NSGlobalDomain ApplePersistence
 
+        salt '*' macdefaults.delete NSGlobalDomain key.with.dots key_separator='.'
+
     domain
         The name of the domain to delete from
 
     key
-        The key of the given domain to delete
+        The key of the given domain to delete.
+        It can be a nested key separated by `key_separator`.
+
+    key_separator
+        The separator to use when splitting the key into a list of keys.
+        If None, the key will not be split (Default).
 
     user
         The user to delete the defaults with
 
     """
-    cmd = f'defaults delete "{domain}" "{key}"'
-    return __salt__["cmd.run_all"](cmd, runas=user, output_loglevel="debug")
+    plist = _load_plist(domain, user=user)
+    if plist is None:
+        return None
+
+    keys = [key] if key_separator is None else key.split(key_separator)
+
+    # Traverse the plist til the penultimate key.
+    # Last key must be handled separately since we
+    # need the parent dictionary to delete that key.
+    target = _traverse_keys(plist, keys[:-1])
+    if target is None:
+        return None
+
+    # Delete the last key if it exists and update defaults
+    last_key = keys[-1]
+    key_in_plist = False
+    if isinstance(target, dict) and last_key in target:
+        key_in_plist = True
+
+    elif (
+        isinstance(target, list)
+        and last_key.isdigit()
+        and -len(target) <= int(last_key) < len(target)
+    ):
+        key_in_plist = True
+        last_key = int(last_key)
+
+    if not key_in_plist:
+        return None
+
+    del target[last_key]
+    return _save_plist(domain, plist, user=user)
+
+
+def cast_value_to_vtype(value, vtype):
+    """
+    Convert the value to the specified vtype.
+    If the value cannot be converted, a ValueError is raised.
+
```

Comment:
Since this entire function is new, we need to add a `..versionadded:: 3008`

Reply:
Done in commit d9243837c283d6659df911e8c2c5a44d7ce125ff